Parse enums from headers & generate code

sheaf opened the pull request:
First pass at trying to automatically generate some of the bindings.

I copied the "enums" section of `imgui.h` into a file, and manually removed the conditional compilation (C preprocessor directives) before passing it on to the parser (I baked in the filepath temporarily, sorry if that messes up you Nix users).

I then perform a Template Haskell splice to generate newtypes and pattern synonyms (I included the splices below).  
I am keeping track of all the docstrings in the backend, but because of Template Haskell limitations I currently don't add Haddock documentation. (This issue is slated to be fixed in `GHC 9.2`.)

While writing these I noticed two oddities:
 - `ImGuiMouseCursor_None` is not `0` but `-1`, so we have to be slightly careful about that (we can't assume `None` always means `0`);
 - `ImGuiColorEditFlags` has some enum patterns that start with double underscores.

A few other details:
 - I derive a `Bits` instance for the flag-like enumerations but not for the others.
 - I removed the `COUNT` patterns. I'm still keeping track of the enum sizes in the backend, so it's just a question of what we want to generate for it. I was thinking perhaps a type family instance
 ```haskell
instance Sized ImGuiCol where
  type Size ImGuiCol = 53
````
 - I didn't remove the `None` patterns as per the observation with `ImGuiMouseCursor_None`.
 - I tried removing the enum name entirely from the patterns, but that causes quite a few clashes even without `None`.


<details>
<summary>Click to show Template Haskell splices.</summary>

```haskell
    newtype ImGuiWindowFlags
      = ImGuiWindowFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiWindowFlags_None :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_None = ImGuiWindowFlags 0
    pattern ImGuiWindowFlags_NoTitleBar :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoTitleBar = ImGuiWindowFlags 1
    pattern ImGuiWindowFlags_NoResize :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoResize = ImGuiWindowFlags 2
    pattern ImGuiWindowFlags_NoMove :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoMove = ImGuiWindowFlags 4
    pattern ImGuiWindowFlags_NoScrollbar :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoScrollbar = ImGuiWindowFlags 8
    pattern ImGuiWindowFlags_NoScrollWithMouse :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoScrollWithMouse = ImGuiWindowFlags 16
    pattern ImGuiWindowFlags_NoCollapse :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoCollapse = ImGuiWindowFlags 32
    pattern ImGuiWindowFlags_AlwaysAutoResize :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_AlwaysAutoResize = ImGuiWindowFlags 64
    pattern ImGuiWindowFlags_NoBackground :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoBackground = ImGuiWindowFlags 128
    pattern ImGuiWindowFlags_NoSavedSettings :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoSavedSettings = ImGuiWindowFlags 256
    pattern ImGuiWindowFlags_NoMouseInputs :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoMouseInputs = ImGuiWindowFlags 512
    pattern ImGuiWindowFlags_MenuBar :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_MenuBar = ImGuiWindowFlags 1024
    pattern ImGuiWindowFlags_HorizontalScrollbar :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_HorizontalScrollbar = ImGuiWindowFlags 2048
    pattern ImGuiWindowFlags_NoFocusOnAppearing :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoFocusOnAppearing = ImGuiWindowFlags 4096
    pattern ImGuiWindowFlags_NoBringToFrontOnFocus :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoBringToFrontOnFocus = ImGuiWindowFlags 8192
    pattern ImGuiWindowFlags_AlwaysVerticalScrollbar ::
              ImGuiWindowFlags
    pattern ImGuiWindowFlags_AlwaysVerticalScrollbar = ImGuiWindowFlags 16384
    pattern ImGuiWindowFlags_AlwaysHorizontalScrollbar ::
              ImGuiWindowFlags
    pattern ImGuiWindowFlags_AlwaysHorizontalScrollbar = ImGuiWindowFlags 32768
    pattern ImGuiWindowFlags_AlwaysUseWindowPadding :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_AlwaysUseWindowPadding = ImGuiWindowFlags 65536
    pattern ImGuiWindowFlags_NoNavInputs :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoNavInputs = ImGuiWindowFlags 262144
    pattern ImGuiWindowFlags_NoNavFocus :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoNavFocus = ImGuiWindowFlags 524288
    pattern ImGuiWindowFlags_UnsavedDocument :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_UnsavedDocument = ImGuiWindowFlags 1048576
    pattern ImGuiWindowFlags_NoNav :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoNav = ImGuiWindowFlags 786432
    pattern ImGuiWindowFlags_NoDecoration :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoDecoration = ImGuiWindowFlags 43
    pattern ImGuiWindowFlags_NoInputs :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NoInputs = ImGuiWindowFlags 786944
    pattern ImGuiWindowFlags_NavFlattened :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_NavFlattened = ImGuiWindowFlags 8388608
    pattern ImGuiWindowFlags_ChildWindow :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_ChildWindow = ImGuiWindowFlags 16777216
    pattern ImGuiWindowFlags_Tooltip :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_Tooltip = ImGuiWindowFlags 33554432
    pattern ImGuiWindowFlags_Popup :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_Popup = ImGuiWindowFlags 67108864
    pattern ImGuiWindowFlags_Modal :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_Modal = ImGuiWindowFlags 134217728
    pattern ImGuiWindowFlags_ChildMenu :: ImGuiWindowFlags
    pattern ImGuiWindowFlags_ChildMenu = ImGuiWindowFlags 268435456
    newtype ImGuiInputTextFlags
      = ImGuiInputTextFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiInputTextFlags_None :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_None = ImGuiInputTextFlags 0
    pattern ImGuiInputTextFlags_CharsDecimal :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CharsDecimal = ImGuiInputTextFlags 1
    pattern ImGuiInputTextFlags_CharsHexadecimal :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CharsHexadecimal = ImGuiInputTextFlags 2
    pattern ImGuiInputTextFlags_CharsUppercase :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CharsUppercase = ImGuiInputTextFlags 4
    pattern ImGuiInputTextFlags_CharsNoBlank :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CharsNoBlank = ImGuiInputTextFlags 8
    pattern ImGuiInputTextFlags_AutoSelectAll :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_AutoSelectAll = ImGuiInputTextFlags 16
    pattern ImGuiInputTextFlags_EnterReturnsTrue :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_EnterReturnsTrue = ImGuiInputTextFlags 32
    pattern ImGuiInputTextFlags_CallbackCompletion ::
              ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackCompletion = ImGuiInputTextFlags 64
    pattern ImGuiInputTextFlags_CallbackHistory :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackHistory = ImGuiInputTextFlags 128
    pattern ImGuiInputTextFlags_CallbackAlways :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackAlways = ImGuiInputTextFlags 256
    pattern ImGuiInputTextFlags_CallbackCharFilter ::
              ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackCharFilter = ImGuiInputTextFlags 512
    pattern ImGuiInputTextFlags_AllowTabInput :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_AllowTabInput = ImGuiInputTextFlags 1024
    pattern ImGuiInputTextFlags_CtrlEnterForNewLine ::
              ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CtrlEnterForNewLine = ImGuiInputTextFlags 2048
    pattern ImGuiInputTextFlags_NoHorizontalScroll ::
              ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_NoHorizontalScroll = ImGuiInputTextFlags 4096
    pattern ImGuiInputTextFlags_AlwaysInsertMode :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_AlwaysInsertMode = ImGuiInputTextFlags 8192
    pattern ImGuiInputTextFlags_ReadOnly :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_ReadOnly = ImGuiInputTextFlags 16384
    pattern ImGuiInputTextFlags_Password :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_Password = ImGuiInputTextFlags 32768
    pattern ImGuiInputTextFlags_NoUndoRedo :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_NoUndoRedo = ImGuiInputTextFlags 65536
    pattern ImGuiInputTextFlags_CharsScientific :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CharsScientific = ImGuiInputTextFlags 131072
    pattern ImGuiInputTextFlags_CallbackResize :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackResize = ImGuiInputTextFlags 262144
    pattern ImGuiInputTextFlags_CallbackEdit :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_CallbackEdit = ImGuiInputTextFlags 524288
    pattern ImGuiInputTextFlags_Multiline :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_Multiline = ImGuiInputTextFlags 1048576
    pattern ImGuiInputTextFlags_NoMarkEdited :: ImGuiInputTextFlags
    pattern ImGuiInputTextFlags_NoMarkEdited = ImGuiInputTextFlags 2097152
    newtype ImGuiTreeNodeFlags
      = ImGuiTreeNodeFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTreeNodeFlags_None :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_None = ImGuiTreeNodeFlags 0
    pattern ImGuiTreeNodeFlags_Selected :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_Selected = ImGuiTreeNodeFlags 1
    pattern ImGuiTreeNodeFlags_Framed :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_Framed = ImGuiTreeNodeFlags 2
    pattern ImGuiTreeNodeFlags_AllowItemOverlap :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_AllowItemOverlap = ImGuiTreeNodeFlags 4
    pattern ImGuiTreeNodeFlags_NoTreePushOnOpen :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_NoTreePushOnOpen = ImGuiTreeNodeFlags 8
    pattern ImGuiTreeNodeFlags_NoAutoOpenOnLog :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_NoAutoOpenOnLog = ImGuiTreeNodeFlags 16
    pattern ImGuiTreeNodeFlags_DefaultOpen :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_DefaultOpen = ImGuiTreeNodeFlags 32
    pattern ImGuiTreeNodeFlags_OpenOnDoubleClick :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_OpenOnDoubleClick = ImGuiTreeNodeFlags 64
    pattern ImGuiTreeNodeFlags_OpenOnArrow :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_OpenOnArrow = ImGuiTreeNodeFlags 128
    pattern ImGuiTreeNodeFlags_Leaf :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_Leaf = ImGuiTreeNodeFlags 256
    pattern ImGuiTreeNodeFlags_Bullet :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_Bullet = ImGuiTreeNodeFlags 512
    pattern ImGuiTreeNodeFlags_FramePadding :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_FramePadding = ImGuiTreeNodeFlags 1024
    pattern ImGuiTreeNodeFlags_SpanAvailWidth :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_SpanAvailWidth = ImGuiTreeNodeFlags 2048
    pattern ImGuiTreeNodeFlags_SpanFullWidth :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_SpanFullWidth = ImGuiTreeNodeFlags 4096
    pattern ImGuiTreeNodeFlags_NavLeftJumpsBackHere ::
              ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_NavLeftJumpsBackHere = ImGuiTreeNodeFlags 8192
    pattern ImGuiTreeNodeFlags_CollapsingHeader :: ImGuiTreeNodeFlags
    pattern ImGuiTreeNodeFlags_CollapsingHeader = ImGuiTreeNodeFlags 26
    newtype ImGuiPopupFlags
      = ImGuiPopupFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiPopupFlags_None :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_None = ImGuiPopupFlags 0
    pattern ImGuiPopupFlags_MouseButtonLeft :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_MouseButtonLeft = ImGuiPopupFlags 0
    pattern ImGuiPopupFlags_MouseButtonRight :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_MouseButtonRight = ImGuiPopupFlags 1
    pattern ImGuiPopupFlags_MouseButtonMiddle :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_MouseButtonMiddle = ImGuiPopupFlags 2
    pattern ImGuiPopupFlags_MouseButtonMask_ :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_MouseButtonMask_ = ImGuiPopupFlags 31
    pattern ImGuiPopupFlags_MouseButtonDefault_ :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_MouseButtonDefault_ = ImGuiPopupFlags 1
    pattern ImGuiPopupFlags_NoOpenOverExistingPopup :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_NoOpenOverExistingPopup = ImGuiPopupFlags 32
    pattern ImGuiPopupFlags_NoOpenOverItems :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_NoOpenOverItems = ImGuiPopupFlags 64
    pattern ImGuiPopupFlags_AnyPopupId :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_AnyPopupId = ImGuiPopupFlags 128
    pattern ImGuiPopupFlags_AnyPopupLevel :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_AnyPopupLevel = ImGuiPopupFlags 256
    pattern ImGuiPopupFlags_AnyPopup :: ImGuiPopupFlags
    pattern ImGuiPopupFlags_AnyPopup = ImGuiPopupFlags 384
    newtype ImGuiSelectableFlags
      = ImGuiSelectableFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiSelectableFlags_None :: ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_None = ImGuiSelectableFlags 0
    pattern ImGuiSelectableFlags_DontClosePopups ::
              ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_DontClosePopups = ImGuiSelectableFlags 1
    pattern ImGuiSelectableFlags_SpanAllColumns :: ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_SpanAllColumns = ImGuiSelectableFlags 2
    pattern ImGuiSelectableFlags_AllowDoubleClick ::
              ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_AllowDoubleClick = ImGuiSelectableFlags 4
    pattern ImGuiSelectableFlags_Disabled :: ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_Disabled = ImGuiSelectableFlags 8
    pattern ImGuiSelectableFlags_AllowItemOverlap ::
              ImGuiSelectableFlags
    pattern ImGuiSelectableFlags_AllowItemOverlap = ImGuiSelectableFlags 16
    newtype ImGuiComboFlags
      = ImGuiComboFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiComboFlags_None :: ImGuiComboFlags
    pattern ImGuiComboFlags_None = ImGuiComboFlags 0
    pattern ImGuiComboFlags_PopupAlignLeft :: ImGuiComboFlags
    pattern ImGuiComboFlags_PopupAlignLeft = ImGuiComboFlags 1
    pattern ImGuiComboFlags_HeightSmall :: ImGuiComboFlags
    pattern ImGuiComboFlags_HeightSmall = ImGuiComboFlags 2
    pattern ImGuiComboFlags_HeightRegular :: ImGuiComboFlags
    pattern ImGuiComboFlags_HeightRegular = ImGuiComboFlags 4
    pattern ImGuiComboFlags_HeightLarge :: ImGuiComboFlags
    pattern ImGuiComboFlags_HeightLarge = ImGuiComboFlags 8
    pattern ImGuiComboFlags_HeightLargest :: ImGuiComboFlags
    pattern ImGuiComboFlags_HeightLargest = ImGuiComboFlags 16
    pattern ImGuiComboFlags_NoArrowButton :: ImGuiComboFlags
    pattern ImGuiComboFlags_NoArrowButton = ImGuiComboFlags 32
    pattern ImGuiComboFlags_NoPreview :: ImGuiComboFlags
    pattern ImGuiComboFlags_NoPreview = ImGuiComboFlags 64
    pattern ImGuiComboFlags_HeightMask_ :: ImGuiComboFlags
    pattern ImGuiComboFlags_HeightMask_ = ImGuiComboFlags 30
    newtype ImGuiTabBarFlags
      = ImGuiTabBarFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTabBarFlags_None :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_None = ImGuiTabBarFlags 0
    pattern ImGuiTabBarFlags_Reorderable :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_Reorderable = ImGuiTabBarFlags 1
    pattern ImGuiTabBarFlags_AutoSelectNewTabs :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_AutoSelectNewTabs = ImGuiTabBarFlags 2
    pattern ImGuiTabBarFlags_TabListPopupButton :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_TabListPopupButton = ImGuiTabBarFlags 4
    pattern ImGuiTabBarFlags_NoCloseWithMiddleMouseButton ::
              ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_NoCloseWithMiddleMouseButton = ImGuiTabBarFlags 8
    pattern ImGuiTabBarFlags_NoTabListScrollingButtons ::
              ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_NoTabListScrollingButtons = ImGuiTabBarFlags 16
    pattern ImGuiTabBarFlags_NoTooltip :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_NoTooltip = ImGuiTabBarFlags 32
    pattern ImGuiTabBarFlags_FittingPolicyResizeDown ::
              ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_FittingPolicyResizeDown = ImGuiTabBarFlags 64
    pattern ImGuiTabBarFlags_FittingPolicyScroll :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_FittingPolicyScroll = ImGuiTabBarFlags 128
    pattern ImGuiTabBarFlags_FittingPolicyMask_ :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_FittingPolicyMask_ = ImGuiTabBarFlags 192
    pattern ImGuiTabBarFlags_FittingPolicyDefault_ :: ImGuiTabBarFlags
    pattern ImGuiTabBarFlags_FittingPolicyDefault_ = ImGuiTabBarFlags 64
    newtype ImGuiTabItemFlags
      = ImGuiTabItemFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTabItemFlags_None :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_None = ImGuiTabItemFlags 0
    pattern ImGuiTabItemFlags_UnsavedDocument :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_UnsavedDocument = ImGuiTabItemFlags 1
    pattern ImGuiTabItemFlags_SetSelected :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_SetSelected = ImGuiTabItemFlags 2
    pattern ImGuiTabItemFlags_NoCloseWithMiddleMouseButton ::
              ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_NoCloseWithMiddleMouseButton = ImGuiTabItemFlags 4
    pattern ImGuiTabItemFlags_NoPushId :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_NoPushId = ImGuiTabItemFlags 8
    pattern ImGuiTabItemFlags_NoTooltip :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_NoTooltip = ImGuiTabItemFlags 16
    pattern ImGuiTabItemFlags_NoReorder :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_NoReorder = ImGuiTabItemFlags 32
    pattern ImGuiTabItemFlags_Leading :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_Leading = ImGuiTabItemFlags 64
    pattern ImGuiTabItemFlags_Trailing :: ImGuiTabItemFlags
    pattern ImGuiTabItemFlags_Trailing = ImGuiTabItemFlags 128
    newtype ImGuiTableFlags
      = ImGuiTableFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTableFlags_None :: ImGuiTableFlags
    pattern ImGuiTableFlags_None = ImGuiTableFlags 0
    pattern ImGuiTableFlags_Resizable :: ImGuiTableFlags
    pattern ImGuiTableFlags_Resizable = ImGuiTableFlags 1
    pattern ImGuiTableFlags_Reorderable :: ImGuiTableFlags
    pattern ImGuiTableFlags_Reorderable = ImGuiTableFlags 2
    pattern ImGuiTableFlags_Hideable :: ImGuiTableFlags
    pattern ImGuiTableFlags_Hideable = ImGuiTableFlags 4
    pattern ImGuiTableFlags_Sortable :: ImGuiTableFlags
    pattern ImGuiTableFlags_Sortable = ImGuiTableFlags 8
    pattern ImGuiTableFlags_NoSavedSettings :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoSavedSettings = ImGuiTableFlags 16
    pattern ImGuiTableFlags_ContextMenuInBody :: ImGuiTableFlags
    pattern ImGuiTableFlags_ContextMenuInBody = ImGuiTableFlags 32
    pattern ImGuiTableFlags_RowBg :: ImGuiTableFlags
    pattern ImGuiTableFlags_RowBg = ImGuiTableFlags 64
    pattern ImGuiTableFlags_BordersInnerH :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersInnerH = ImGuiTableFlags 128
    pattern ImGuiTableFlags_BordersOuterH :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersOuterH = ImGuiTableFlags 256
    pattern ImGuiTableFlags_BordersInnerV :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersInnerV = ImGuiTableFlags 512
    pattern ImGuiTableFlags_BordersOuterV :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersOuterV = ImGuiTableFlags 1024
    pattern ImGuiTableFlags_BordersH :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersH = ImGuiTableFlags 384
    pattern ImGuiTableFlags_BordersV :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersV = ImGuiTableFlags 1536
    pattern ImGuiTableFlags_BordersInner :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersInner = ImGuiTableFlags 640
    pattern ImGuiTableFlags_BordersOuter :: ImGuiTableFlags
    pattern ImGuiTableFlags_BordersOuter = ImGuiTableFlags 1280
    pattern ImGuiTableFlags_Borders :: ImGuiTableFlags
    pattern ImGuiTableFlags_Borders = ImGuiTableFlags 1920
    pattern ImGuiTableFlags_NoBordersInBody :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoBordersInBody = ImGuiTableFlags 2048
    pattern ImGuiTableFlags_NoBordersInBodyUntilResize ::
              ImGuiTableFlags
    pattern ImGuiTableFlags_NoBordersInBodyUntilResize = ImGuiTableFlags 4096
    pattern ImGuiTableFlags_SizingFixedFit :: ImGuiTableFlags
    pattern ImGuiTableFlags_SizingFixedFit = ImGuiTableFlags 8192
    pattern ImGuiTableFlags_SizingFixedSame :: ImGuiTableFlags
    pattern ImGuiTableFlags_SizingFixedSame = ImGuiTableFlags 16384
    pattern ImGuiTableFlags_SizingStretchProp :: ImGuiTableFlags
    pattern ImGuiTableFlags_SizingStretchProp = ImGuiTableFlags 24576
    pattern ImGuiTableFlags_SizingStretchSame :: ImGuiTableFlags
    pattern ImGuiTableFlags_SizingStretchSame = ImGuiTableFlags 32768
    pattern ImGuiTableFlags_NoHostExtendX :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoHostExtendX = ImGuiTableFlags 65536
    pattern ImGuiTableFlags_NoHostExtendY :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoHostExtendY = ImGuiTableFlags 131072
    pattern ImGuiTableFlags_NoKeepColumnsVisible :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoKeepColumnsVisible = ImGuiTableFlags 262144
    pattern ImGuiTableFlags_PreciseWidths :: ImGuiTableFlags
    pattern ImGuiTableFlags_PreciseWidths = ImGuiTableFlags 524288
    pattern ImGuiTableFlags_NoClip :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoClip = ImGuiTableFlags 1048576
    pattern ImGuiTableFlags_PadOuterX :: ImGuiTableFlags
    pattern ImGuiTableFlags_PadOuterX = ImGuiTableFlags 2097152
    pattern ImGuiTableFlags_NoPadOuterX :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoPadOuterX = ImGuiTableFlags 4194304
    pattern ImGuiTableFlags_NoPadInnerX :: ImGuiTableFlags
    pattern ImGuiTableFlags_NoPadInnerX = ImGuiTableFlags 8388608
    pattern ImGuiTableFlags_ScrollX :: ImGuiTableFlags
    pattern ImGuiTableFlags_ScrollX = ImGuiTableFlags 16777216
    pattern ImGuiTableFlags_ScrollY :: ImGuiTableFlags
    pattern ImGuiTableFlags_ScrollY = ImGuiTableFlags 33554432
    pattern ImGuiTableFlags_SortMulti :: ImGuiTableFlags
    pattern ImGuiTableFlags_SortMulti = ImGuiTableFlags 67108864
    pattern ImGuiTableFlags_SortTristate :: ImGuiTableFlags
    pattern ImGuiTableFlags_SortTristate = ImGuiTableFlags 134217728
    pattern ImGuiTableFlags_SizingMask_ :: ImGuiTableFlags
    pattern ImGuiTableFlags_SizingMask_ = ImGuiTableFlags 81920
    newtype ImGuiTableColumnFlags
      = ImGuiTableColumnFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTableColumnFlags_None :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_None = ImGuiTableColumnFlags 0
    pattern ImGuiTableColumnFlags_DefaultHide :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_DefaultHide = ImGuiTableColumnFlags 1
    pattern ImGuiTableColumnFlags_DefaultSort :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_DefaultSort = ImGuiTableColumnFlags 2
    pattern ImGuiTableColumnFlags_WidthStretch :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_WidthStretch = ImGuiTableColumnFlags 4
    pattern ImGuiTableColumnFlags_WidthFixed :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_WidthFixed = ImGuiTableColumnFlags 8
    pattern ImGuiTableColumnFlags_NoResize :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoResize = ImGuiTableColumnFlags 16
    pattern ImGuiTableColumnFlags_NoReorder :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoReorder = ImGuiTableColumnFlags 32
    pattern ImGuiTableColumnFlags_NoHide :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoHide = ImGuiTableColumnFlags 64
    pattern ImGuiTableColumnFlags_NoClip :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoClip = ImGuiTableColumnFlags 128
    pattern ImGuiTableColumnFlags_NoSort :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoSort = ImGuiTableColumnFlags 256
    pattern ImGuiTableColumnFlags_NoSortAscending ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoSortAscending = ImGuiTableColumnFlags 512
    pattern ImGuiTableColumnFlags_NoSortDescending ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoSortDescending = ImGuiTableColumnFlags 1024
    pattern ImGuiTableColumnFlags_NoHeaderWidth ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoHeaderWidth = ImGuiTableColumnFlags 2048
    pattern ImGuiTableColumnFlags_PreferSortAscending ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_PreferSortAscending = ImGuiTableColumnFlags 4096
    pattern ImGuiTableColumnFlags_PreferSortDescending ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_PreferSortDescending = ImGuiTableColumnFlags 8192
    pattern ImGuiTableColumnFlags_IndentEnable :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IndentEnable = ImGuiTableColumnFlags 16384
    pattern ImGuiTableColumnFlags_IndentDisable ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IndentDisable = ImGuiTableColumnFlags 32768
    pattern ImGuiTableColumnFlags_IsEnabled :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IsEnabled = ImGuiTableColumnFlags 1048576
    pattern ImGuiTableColumnFlags_IsVisible :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IsVisible = ImGuiTableColumnFlags 2097152
    pattern ImGuiTableColumnFlags_IsSorted :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IsSorted = ImGuiTableColumnFlags 4194304
    pattern ImGuiTableColumnFlags_IsHovered :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IsHovered = ImGuiTableColumnFlags 8388608
    pattern ImGuiTableColumnFlags_WidthMask_ :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_WidthMask_ = ImGuiTableColumnFlags 12
    pattern ImGuiTableColumnFlags_IndentMask_ :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_IndentMask_ = ImGuiTableColumnFlags 49152
    pattern ImGuiTableColumnFlags_StatusMask_ :: ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_StatusMask_ = ImGuiTableColumnFlags 15728640
    pattern ImGuiTableColumnFlags_NoDirectResize_ ::
              ImGuiTableColumnFlags
    pattern ImGuiTableColumnFlags_NoDirectResize_ = ImGuiTableColumnFlags 1073741824
    newtype ImGuiTableRowFlags
      = ImGuiTableRowFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiTableRowFlags_None :: ImGuiTableRowFlags
    pattern ImGuiTableRowFlags_None = ImGuiTableRowFlags 0
    pattern ImGuiTableRowFlags_Headers :: ImGuiTableRowFlags
    pattern ImGuiTableRowFlags_Headers = ImGuiTableRowFlags 1
    newtype ImGuiTableBgTarget
      = ImGuiTableBgTarget Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiTableBgTarget_None :: ImGuiTableBgTarget
    pattern ImGuiTableBgTarget_None = ImGuiTableBgTarget 0
    pattern ImGuiTableBgTarget_RowBg0 :: ImGuiTableBgTarget
    pattern ImGuiTableBgTarget_RowBg0 = ImGuiTableBgTarget 1
    pattern ImGuiTableBgTarget_RowBg1 :: ImGuiTableBgTarget
    pattern ImGuiTableBgTarget_RowBg1 = ImGuiTableBgTarget 2
    pattern ImGuiTableBgTarget_CellBg :: ImGuiTableBgTarget
    pattern ImGuiTableBgTarget_CellBg = ImGuiTableBgTarget 3
    newtype ImGuiFocusedFlags
      = ImGuiFocusedFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiFocusedFlags_None :: ImGuiFocusedFlags
    pattern ImGuiFocusedFlags_None = ImGuiFocusedFlags 0
    pattern ImGuiFocusedFlags_ChildWindows :: ImGuiFocusedFlags
    pattern ImGuiFocusedFlags_ChildWindows = ImGuiFocusedFlags 1
    pattern ImGuiFocusedFlags_RootWindow :: ImGuiFocusedFlags
    pattern ImGuiFocusedFlags_RootWindow = ImGuiFocusedFlags 2
    pattern ImGuiFocusedFlags_AnyWindow :: ImGuiFocusedFlags
    pattern ImGuiFocusedFlags_AnyWindow = ImGuiFocusedFlags 4
    pattern ImGuiFocusedFlags_RootAndChildWindows :: ImGuiFocusedFlags
    pattern ImGuiFocusedFlags_RootAndChildWindows = ImGuiFocusedFlags 3
    newtype ImGuiHoveredFlags
      = ImGuiHoveredFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiHoveredFlags_None :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_None = ImGuiHoveredFlags 0
    pattern ImGuiHoveredFlags_ChildWindows :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_ChildWindows = ImGuiHoveredFlags 1
    pattern ImGuiHoveredFlags_RootWindow :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_RootWindow = ImGuiHoveredFlags 2
    pattern ImGuiHoveredFlags_AnyWindow :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_AnyWindow = ImGuiHoveredFlags 4
    pattern ImGuiHoveredFlags_AllowWhenBlockedByPopup ::
              ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_AllowWhenBlockedByPopup = ImGuiHoveredFlags 8
    pattern ImGuiHoveredFlags_AllowWhenBlockedByActiveItem ::
              ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_AllowWhenBlockedByActiveItem = ImGuiHoveredFlags 32
    pattern ImGuiHoveredFlags_AllowWhenOverlapped :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_AllowWhenOverlapped = ImGuiHoveredFlags 64
    pattern ImGuiHoveredFlags_AllowWhenDisabled :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_AllowWhenDisabled = ImGuiHoveredFlags 128
    pattern ImGuiHoveredFlags_RectOnly :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_RectOnly = ImGuiHoveredFlags 104
    pattern ImGuiHoveredFlags_RootAndChildWindows :: ImGuiHoveredFlags
    pattern ImGuiHoveredFlags_RootAndChildWindows = ImGuiHoveredFlags 3
    newtype ImGuiDragDropFlags
      = ImGuiDragDropFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiDragDropFlags_None :: ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_None = ImGuiDragDropFlags 0
    pattern ImGuiDragDropFlags_SourceNoPreviewTooltip ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceNoPreviewTooltip = ImGuiDragDropFlags 1
    pattern ImGuiDragDropFlags_SourceNoDisableHover ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceNoDisableHover = ImGuiDragDropFlags 2
    pattern ImGuiDragDropFlags_SourceNoHoldToOpenOthers ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceNoHoldToOpenOthers = ImGuiDragDropFlags 4
    pattern ImGuiDragDropFlags_SourceAllowNullID :: ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceAllowNullID = ImGuiDragDropFlags 8
    pattern ImGuiDragDropFlags_SourceExtern :: ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceExtern = ImGuiDragDropFlags 16
    pattern ImGuiDragDropFlags_SourceAutoExpirePayload ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_SourceAutoExpirePayload = ImGuiDragDropFlags 32
    pattern ImGuiDragDropFlags_AcceptBeforeDelivery ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_AcceptBeforeDelivery = ImGuiDragDropFlags 1024
    pattern ImGuiDragDropFlags_AcceptNoDrawDefaultRect ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_AcceptNoDrawDefaultRect = ImGuiDragDropFlags 2048
    pattern ImGuiDragDropFlags_AcceptNoPreviewTooltip ::
              ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_AcceptNoPreviewTooltip = ImGuiDragDropFlags 4096
    pattern ImGuiDragDropFlags_AcceptPeekOnly :: ImGuiDragDropFlags
    pattern ImGuiDragDropFlags_AcceptPeekOnly = ImGuiDragDropFlags 3072
    newtype ImGuiDataType
      = ImGuiDataType Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiDataType_S8 :: ImGuiDataType
    pattern ImGuiDataType_S8 = ImGuiDataType 0
    pattern ImGuiDataType_U8 :: ImGuiDataType
    pattern ImGuiDataType_U8 = ImGuiDataType 1
    pattern ImGuiDataType_S16 :: ImGuiDataType
    pattern ImGuiDataType_S16 = ImGuiDataType 2
    pattern ImGuiDataType_U16 :: ImGuiDataType
    pattern ImGuiDataType_U16 = ImGuiDataType 3
    pattern ImGuiDataType_S32 :: ImGuiDataType
    pattern ImGuiDataType_S32 = ImGuiDataType 4
    pattern ImGuiDataType_U32 :: ImGuiDataType
    pattern ImGuiDataType_U32 = ImGuiDataType 5
    pattern ImGuiDataType_S64 :: ImGuiDataType
    pattern ImGuiDataType_S64 = ImGuiDataType 6
    pattern ImGuiDataType_U64 :: ImGuiDataType
    pattern ImGuiDataType_U64 = ImGuiDataType 7
    pattern ImGuiDataType_Float :: ImGuiDataType
    pattern ImGuiDataType_Float = ImGuiDataType 8
    pattern ImGuiDataType_Double :: ImGuiDataType
    pattern ImGuiDataType_Double = ImGuiDataType 9
    newtype ImGuiDir
      = ImGuiDir Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiDir_None :: ImGuiDir
    pattern ImGuiDir_None = ImGuiDir (-1)
    pattern ImGuiDir_Left :: ImGuiDir
    pattern ImGuiDir_Left = ImGuiDir 0
    pattern ImGuiDir_Right :: ImGuiDir
    pattern ImGuiDir_Right = ImGuiDir 1
    pattern ImGuiDir_Up :: ImGuiDir
    pattern ImGuiDir_Up = ImGuiDir 2
    pattern ImGuiDir_Down :: ImGuiDir
    pattern ImGuiDir_Down = ImGuiDir 3
    newtype ImGuiSortDirection
      = ImGuiSortDirection Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiSortDirection_None :: ImGuiSortDirection
    pattern ImGuiSortDirection_None = ImGuiSortDirection 0
    pattern ImGuiSortDirection_Ascending :: ImGuiSortDirection
    pattern ImGuiSortDirection_Ascending = ImGuiSortDirection 1
    pattern ImGuiSortDirection_Descending :: ImGuiSortDirection
    pattern ImGuiSortDirection_Descending = ImGuiSortDirection 2
    newtype ImGuiKey
      = ImGuiKey Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiKey_Tab :: ImGuiKey
    pattern ImGuiKey_Tab = ImGuiKey 0
    pattern ImGuiKey_LeftArrow :: ImGuiKey
    pattern ImGuiKey_LeftArrow = ImGuiKey 1
    pattern ImGuiKey_RightArrow :: ImGuiKey
    pattern ImGuiKey_RightArrow = ImGuiKey 2
    pattern ImGuiKey_UpArrow :: ImGuiKey
    pattern ImGuiKey_UpArrow = ImGuiKey 3
    pattern ImGuiKey_DownArrow :: ImGuiKey
    pattern ImGuiKey_DownArrow = ImGuiKey 4
    pattern ImGuiKey_PageUp :: ImGuiKey
    pattern ImGuiKey_PageUp = ImGuiKey 5
    pattern ImGuiKey_PageDown :: ImGuiKey
    pattern ImGuiKey_PageDown = ImGuiKey 6
    pattern ImGuiKey_Home :: ImGuiKey
    pattern ImGuiKey_Home = ImGuiKey 7
    pattern ImGuiKey_End :: ImGuiKey
    pattern ImGuiKey_End = ImGuiKey 8
    pattern ImGuiKey_Insert :: ImGuiKey
    pattern ImGuiKey_Insert = ImGuiKey 9
    pattern ImGuiKey_Delete :: ImGuiKey
    pattern ImGuiKey_Delete = ImGuiKey 10
    pattern ImGuiKey_Backspace :: ImGuiKey
    pattern ImGuiKey_Backspace = ImGuiKey 11
    pattern ImGuiKey_Space :: ImGuiKey
    pattern ImGuiKey_Space = ImGuiKey 12
    pattern ImGuiKey_Enter :: ImGuiKey
    pattern ImGuiKey_Enter = ImGuiKey 13
    pattern ImGuiKey_Escape :: ImGuiKey
    pattern ImGuiKey_Escape = ImGuiKey 14
    pattern ImGuiKey_KeyPadEnter :: ImGuiKey
    pattern ImGuiKey_KeyPadEnter = ImGuiKey 15
    pattern ImGuiKey_A :: ImGuiKey
    pattern ImGuiKey_A = ImGuiKey 16
    pattern ImGuiKey_C :: ImGuiKey
    pattern ImGuiKey_C = ImGuiKey 17
    pattern ImGuiKey_V :: ImGuiKey
    pattern ImGuiKey_V = ImGuiKey 18
    pattern ImGuiKey_X :: ImGuiKey
    pattern ImGuiKey_X = ImGuiKey 19
    pattern ImGuiKey_Y :: ImGuiKey
    pattern ImGuiKey_Y = ImGuiKey 20
    pattern ImGuiKey_Z :: ImGuiKey
    pattern ImGuiKey_Z = ImGuiKey 21
    newtype ImGuiKeyModFlags
      = ImGuiKeyModFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiKeyModFlags_None :: ImGuiKeyModFlags
    pattern ImGuiKeyModFlags_None = ImGuiKeyModFlags 0
    pattern ImGuiKeyModFlags_Ctrl :: ImGuiKeyModFlags
    pattern ImGuiKeyModFlags_Ctrl = ImGuiKeyModFlags 1
    pattern ImGuiKeyModFlags_Shift :: ImGuiKeyModFlags
    pattern ImGuiKeyModFlags_Shift = ImGuiKeyModFlags 2
    pattern ImGuiKeyModFlags_Alt :: ImGuiKeyModFlags
    pattern ImGuiKeyModFlags_Alt = ImGuiKeyModFlags 4
    pattern ImGuiKeyModFlags_Super :: ImGuiKeyModFlags
    pattern ImGuiKeyModFlags_Super = ImGuiKeyModFlags 8
    newtype ImGuiNavInput
      = ImGuiNavInput Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiNavInput_Activate :: ImGuiNavInput
    pattern ImGuiNavInput_Activate = ImGuiNavInput 0
    pattern ImGuiNavInput_Cancel :: ImGuiNavInput
    pattern ImGuiNavInput_Cancel = ImGuiNavInput 1
    pattern ImGuiNavInput_Input :: ImGuiNavInput
    pattern ImGuiNavInput_Input = ImGuiNavInput 2
    pattern ImGuiNavInput_Menu :: ImGuiNavInput
    pattern ImGuiNavInput_Menu = ImGuiNavInput 3
    pattern ImGuiNavInput_DpadLeft :: ImGuiNavInput
    pattern ImGuiNavInput_DpadLeft = ImGuiNavInput 4
    pattern ImGuiNavInput_DpadRight :: ImGuiNavInput
    pattern ImGuiNavInput_DpadRight = ImGuiNavInput 5
    pattern ImGuiNavInput_DpadUp :: ImGuiNavInput
    pattern ImGuiNavInput_DpadUp = ImGuiNavInput 6
    pattern ImGuiNavInput_DpadDown :: ImGuiNavInput
    pattern ImGuiNavInput_DpadDown = ImGuiNavInput 7
    pattern ImGuiNavInput_LStickLeft :: ImGuiNavInput
    pattern ImGuiNavInput_LStickLeft = ImGuiNavInput 8
    pattern ImGuiNavInput_LStickRight :: ImGuiNavInput
    pattern ImGuiNavInput_LStickRight = ImGuiNavInput 9
    pattern ImGuiNavInput_LStickUp :: ImGuiNavInput
    pattern ImGuiNavInput_LStickUp = ImGuiNavInput 10
    pattern ImGuiNavInput_LStickDown :: ImGuiNavInput
    pattern ImGuiNavInput_LStickDown = ImGuiNavInput 11
    pattern ImGuiNavInput_FocusPrev :: ImGuiNavInput
    pattern ImGuiNavInput_FocusPrev = ImGuiNavInput 12
    pattern ImGuiNavInput_FocusNext :: ImGuiNavInput
    pattern ImGuiNavInput_FocusNext = ImGuiNavInput 13
    pattern ImGuiNavInput_TweakSlow :: ImGuiNavInput
    pattern ImGuiNavInput_TweakSlow = ImGuiNavInput 14
    pattern ImGuiNavInput_TweakFast :: ImGuiNavInput
    pattern ImGuiNavInput_TweakFast = ImGuiNavInput 15
    pattern ImGuiNavInput_KeyMenu_ :: ImGuiNavInput
    pattern ImGuiNavInput_KeyMenu_ = ImGuiNavInput 16
    pattern ImGuiNavInput_KeyLeft_ :: ImGuiNavInput
    pattern ImGuiNavInput_KeyLeft_ = ImGuiNavInput 17
    pattern ImGuiNavInput_KeyRight_ :: ImGuiNavInput
    pattern ImGuiNavInput_KeyRight_ = ImGuiNavInput 18
    pattern ImGuiNavInput_KeyUp_ :: ImGuiNavInput
    pattern ImGuiNavInput_KeyUp_ = ImGuiNavInput 19
    pattern ImGuiNavInput_KeyDown_ :: ImGuiNavInput
    pattern ImGuiNavInput_KeyDown_ = ImGuiNavInput 20
    pattern ImGuiNavInput_InternalStart_ :: ImGuiNavInput
    pattern ImGuiNavInput_InternalStart_ = ImGuiNavInput 16
    newtype ImGuiConfigFlags
      = ImGuiConfigFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiConfigFlags_None :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_None = ImGuiConfigFlags 0
    pattern ImGuiConfigFlags_NavEnableKeyboard :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NavEnableKeyboard = ImGuiConfigFlags 1
    pattern ImGuiConfigFlags_NavEnableGamepad :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NavEnableGamepad = ImGuiConfigFlags 2
    pattern ImGuiConfigFlags_NavEnableSetMousePos :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NavEnableSetMousePos = ImGuiConfigFlags 4
    pattern ImGuiConfigFlags_NavNoCaptureKeyboard :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NavNoCaptureKeyboard = ImGuiConfigFlags 8
    pattern ImGuiConfigFlags_NoMouse :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NoMouse = ImGuiConfigFlags 16
    pattern ImGuiConfigFlags_NoMouseCursorChange :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_NoMouseCursorChange = ImGuiConfigFlags 32
    pattern ImGuiConfigFlags_IsSRGB :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_IsSRGB = ImGuiConfigFlags 1048576
    pattern ImGuiConfigFlags_IsTouchScreen :: ImGuiConfigFlags
    pattern ImGuiConfigFlags_IsTouchScreen = ImGuiConfigFlags 2097152
    newtype ImGuiBackendFlags
      = ImGuiBackendFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiBackendFlags_None :: ImGuiBackendFlags
    pattern ImGuiBackendFlags_None = ImGuiBackendFlags 0
    pattern ImGuiBackendFlags_HasGamepad :: ImGuiBackendFlags
    pattern ImGuiBackendFlags_HasGamepad = ImGuiBackendFlags 1
    pattern ImGuiBackendFlags_HasMouseCursors :: ImGuiBackendFlags
    pattern ImGuiBackendFlags_HasMouseCursors = ImGuiBackendFlags 2
    pattern ImGuiBackendFlags_HasSetMousePos :: ImGuiBackendFlags
    pattern ImGuiBackendFlags_HasSetMousePos = ImGuiBackendFlags 4
    pattern ImGuiBackendFlags_RendererHasVtxOffset :: ImGuiBackendFlags
    pattern ImGuiBackendFlags_RendererHasVtxOffset = ImGuiBackendFlags 8
    newtype ImGuiCol
      = ImGuiCol Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiCol_Text :: ImGuiCol
    pattern ImGuiCol_Text = ImGuiCol 0
    pattern ImGuiCol_TextDisabled :: ImGuiCol
    pattern ImGuiCol_TextDisabled = ImGuiCol 1
    pattern ImGuiCol_WindowBg :: ImGuiCol
    pattern ImGuiCol_WindowBg = ImGuiCol 2
    pattern ImGuiCol_ChildBg :: ImGuiCol
    pattern ImGuiCol_ChildBg = ImGuiCol 3
    pattern ImGuiCol_PopupBg :: ImGuiCol
    pattern ImGuiCol_PopupBg = ImGuiCol 4
    pattern ImGuiCol_Border :: ImGuiCol
    pattern ImGuiCol_Border = ImGuiCol 5
    pattern ImGuiCol_BorderShadow :: ImGuiCol
    pattern ImGuiCol_BorderShadow = ImGuiCol 6
    pattern ImGuiCol_FrameBg :: ImGuiCol
    pattern ImGuiCol_FrameBg = ImGuiCol 7
    pattern ImGuiCol_FrameBgHovered :: ImGuiCol
    pattern ImGuiCol_FrameBgHovered = ImGuiCol 8
    pattern ImGuiCol_FrameBgActive :: ImGuiCol
    pattern ImGuiCol_FrameBgActive = ImGuiCol 9
    pattern ImGuiCol_TitleBg :: ImGuiCol
    pattern ImGuiCol_TitleBg = ImGuiCol 10
    pattern ImGuiCol_TitleBgActive :: ImGuiCol
    pattern ImGuiCol_TitleBgActive = ImGuiCol 11
    pattern ImGuiCol_TitleBgCollapsed :: ImGuiCol
    pattern ImGuiCol_TitleBgCollapsed = ImGuiCol 12
    pattern ImGuiCol_MenuBarBg :: ImGuiCol
    pattern ImGuiCol_MenuBarBg = ImGuiCol 13
    pattern ImGuiCol_ScrollbarBg :: ImGuiCol
    pattern ImGuiCol_ScrollbarBg = ImGuiCol 14
    pattern ImGuiCol_ScrollbarGrab :: ImGuiCol
    pattern ImGuiCol_ScrollbarGrab = ImGuiCol 15
    pattern ImGuiCol_ScrollbarGrabHovered :: ImGuiCol
    pattern ImGuiCol_ScrollbarGrabHovered = ImGuiCol 16
    pattern ImGuiCol_ScrollbarGrabActive :: ImGuiCol
    pattern ImGuiCol_ScrollbarGrabActive = ImGuiCol 17
    pattern ImGuiCol_CheckMark :: ImGuiCol
    pattern ImGuiCol_CheckMark = ImGuiCol 18
    pattern ImGuiCol_SliderGrab :: ImGuiCol
    pattern ImGuiCol_SliderGrab = ImGuiCol 19
    pattern ImGuiCol_SliderGrabActive :: ImGuiCol
    pattern ImGuiCol_SliderGrabActive = ImGuiCol 20
    pattern ImGuiCol_Button :: ImGuiCol
    pattern ImGuiCol_Button = ImGuiCol 21
    pattern ImGuiCol_ButtonHovered :: ImGuiCol
    pattern ImGuiCol_ButtonHovered = ImGuiCol 22
    pattern ImGuiCol_ButtonActive :: ImGuiCol
    pattern ImGuiCol_ButtonActive = ImGuiCol 23
    pattern ImGuiCol_Header :: ImGuiCol
    pattern ImGuiCol_Header = ImGuiCol 24
    pattern ImGuiCol_HeaderHovered :: ImGuiCol
    pattern ImGuiCol_HeaderHovered = ImGuiCol 25
    pattern ImGuiCol_HeaderActive :: ImGuiCol
    pattern ImGuiCol_HeaderActive = ImGuiCol 26
    pattern ImGuiCol_Separator :: ImGuiCol
    pattern ImGuiCol_Separator = ImGuiCol 27
    pattern ImGuiCol_SeparatorHovered :: ImGuiCol
    pattern ImGuiCol_SeparatorHovered = ImGuiCol 28
    pattern ImGuiCol_SeparatorActive :: ImGuiCol
    pattern ImGuiCol_SeparatorActive = ImGuiCol 29
    pattern ImGuiCol_ResizeGrip :: ImGuiCol
    pattern ImGuiCol_ResizeGrip = ImGuiCol 30
    pattern ImGuiCol_ResizeGripHovered :: ImGuiCol
    pattern ImGuiCol_ResizeGripHovered = ImGuiCol 31
    pattern ImGuiCol_ResizeGripActive :: ImGuiCol
    pattern ImGuiCol_ResizeGripActive = ImGuiCol 32
    pattern ImGuiCol_Tab :: ImGuiCol
    pattern ImGuiCol_Tab = ImGuiCol 33
    pattern ImGuiCol_TabHovered :: ImGuiCol
    pattern ImGuiCol_TabHovered = ImGuiCol 34
    pattern ImGuiCol_TabActive :: ImGuiCol
    pattern ImGuiCol_TabActive = ImGuiCol 35
    pattern ImGuiCol_TabUnfocused :: ImGuiCol
    pattern ImGuiCol_TabUnfocused = ImGuiCol 36
    pattern ImGuiCol_TabUnfocusedActive :: ImGuiCol
    pattern ImGuiCol_TabUnfocusedActive = ImGuiCol 37
    pattern ImGuiCol_PlotLines :: ImGuiCol
    pattern ImGuiCol_PlotLines = ImGuiCol 38
    pattern ImGuiCol_PlotLinesHovered :: ImGuiCol
    pattern ImGuiCol_PlotLinesHovered = ImGuiCol 39
    pattern ImGuiCol_PlotHistogram :: ImGuiCol
    pattern ImGuiCol_PlotHistogram = ImGuiCol 40
    pattern ImGuiCol_PlotHistogramHovered :: ImGuiCol
    pattern ImGuiCol_PlotHistogramHovered = ImGuiCol 41
    pattern ImGuiCol_TableHeaderBg :: ImGuiCol
    pattern ImGuiCol_TableHeaderBg = ImGuiCol 42
    pattern ImGuiCol_TableBorderStrong :: ImGuiCol
    pattern ImGuiCol_TableBorderStrong = ImGuiCol 43
    pattern ImGuiCol_TableBorderLight :: ImGuiCol
    pattern ImGuiCol_TableBorderLight = ImGuiCol 44
    pattern ImGuiCol_TableRowBg :: ImGuiCol
    pattern ImGuiCol_TableRowBg = ImGuiCol 45
    pattern ImGuiCol_TableRowBgAlt :: ImGuiCol
    pattern ImGuiCol_TableRowBgAlt = ImGuiCol 46
    pattern ImGuiCol_TextSelectedBg :: ImGuiCol
    pattern ImGuiCol_TextSelectedBg = ImGuiCol 47
    pattern ImGuiCol_DragDropTarget :: ImGuiCol
    pattern ImGuiCol_DragDropTarget = ImGuiCol 48
    pattern ImGuiCol_NavHighlight :: ImGuiCol
    pattern ImGuiCol_NavHighlight = ImGuiCol 49
    pattern ImGuiCol_NavWindowingHighlight :: ImGuiCol
    pattern ImGuiCol_NavWindowingHighlight = ImGuiCol 50
    pattern ImGuiCol_NavWindowingDimBg :: ImGuiCol
    pattern ImGuiCol_NavWindowingDimBg = ImGuiCol 51
    pattern ImGuiCol_ModalWindowDimBg :: ImGuiCol
    pattern ImGuiCol_ModalWindowDimBg = ImGuiCol 52
    newtype ImGuiStyleVar
      = ImGuiStyleVar Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiStyleVar_Alpha :: ImGuiStyleVar
    pattern ImGuiStyleVar_Alpha = ImGuiStyleVar 0
    pattern ImGuiStyleVar_WindowPadding :: ImGuiStyleVar
    pattern ImGuiStyleVar_WindowPadding = ImGuiStyleVar 1
    pattern ImGuiStyleVar_WindowRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_WindowRounding = ImGuiStyleVar 2
    pattern ImGuiStyleVar_WindowBorderSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_WindowBorderSize = ImGuiStyleVar 3
    pattern ImGuiStyleVar_WindowMinSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_WindowMinSize = ImGuiStyleVar 4
    pattern ImGuiStyleVar_WindowTitleAlign :: ImGuiStyleVar
    pattern ImGuiStyleVar_WindowTitleAlign = ImGuiStyleVar 5
    pattern ImGuiStyleVar_ChildRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_ChildRounding = ImGuiStyleVar 6
    pattern ImGuiStyleVar_ChildBorderSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_ChildBorderSize = ImGuiStyleVar 7
    pattern ImGuiStyleVar_PopupRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_PopupRounding = ImGuiStyleVar 8
    pattern ImGuiStyleVar_PopupBorderSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_PopupBorderSize = ImGuiStyleVar 9
    pattern ImGuiStyleVar_FramePadding :: ImGuiStyleVar
    pattern ImGuiStyleVar_FramePadding = ImGuiStyleVar 10
    pattern ImGuiStyleVar_FrameRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_FrameRounding = ImGuiStyleVar 11
    pattern ImGuiStyleVar_FrameBorderSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_FrameBorderSize = ImGuiStyleVar 12
    pattern ImGuiStyleVar_ItemSpacing :: ImGuiStyleVar
    pattern ImGuiStyleVar_ItemSpacing = ImGuiStyleVar 13
    pattern ImGuiStyleVar_ItemInnerSpacing :: ImGuiStyleVar
    pattern ImGuiStyleVar_ItemInnerSpacing = ImGuiStyleVar 14
    pattern ImGuiStyleVar_IndentSpacing :: ImGuiStyleVar
    pattern ImGuiStyleVar_IndentSpacing = ImGuiStyleVar 15
    pattern ImGuiStyleVar_CellPadding :: ImGuiStyleVar
    pattern ImGuiStyleVar_CellPadding = ImGuiStyleVar 16
    pattern ImGuiStyleVar_ScrollbarSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_ScrollbarSize = ImGuiStyleVar 17
    pattern ImGuiStyleVar_ScrollbarRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_ScrollbarRounding = ImGuiStyleVar 18
    pattern ImGuiStyleVar_GrabMinSize :: ImGuiStyleVar
    pattern ImGuiStyleVar_GrabMinSize = ImGuiStyleVar 19
    pattern ImGuiStyleVar_GrabRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_GrabRounding = ImGuiStyleVar 20
    pattern ImGuiStyleVar_TabRounding :: ImGuiStyleVar
    pattern ImGuiStyleVar_TabRounding = ImGuiStyleVar 21
    pattern ImGuiStyleVar_ButtonTextAlign :: ImGuiStyleVar
    pattern ImGuiStyleVar_ButtonTextAlign = ImGuiStyleVar 22
    pattern ImGuiStyleVar_SelectableTextAlign :: ImGuiStyleVar
    pattern ImGuiStyleVar_SelectableTextAlign = ImGuiStyleVar 23
    newtype ImGuiButtonFlags
      = ImGuiButtonFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiButtonFlags_None :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_None = ImGuiButtonFlags 0
    pattern ImGuiButtonFlags_MouseButtonLeft :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_MouseButtonLeft = ImGuiButtonFlags 1
    pattern ImGuiButtonFlags_MouseButtonRight :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_MouseButtonRight = ImGuiButtonFlags 2
    pattern ImGuiButtonFlags_MouseButtonMiddle :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_MouseButtonMiddle = ImGuiButtonFlags 4
    pattern ImGuiButtonFlags_MouseButtonMask_ :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_MouseButtonMask_ = ImGuiButtonFlags 7
    pattern ImGuiButtonFlags_MouseButtonDefault_ :: ImGuiButtonFlags
    pattern ImGuiButtonFlags_MouseButtonDefault_ = ImGuiButtonFlags 1
    newtype ImGuiColorEditFlags
      = ImGuiColorEditFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiColorEditFlags_None :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_None = ImGuiColorEditFlags 0
    pattern ImGuiColorEditFlags_NoAlpha :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoAlpha = ImGuiColorEditFlags 2
    pattern ImGuiColorEditFlags_NoPicker :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoPicker = ImGuiColorEditFlags 4
    pattern ImGuiColorEditFlags_NoOptions :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoOptions = ImGuiColorEditFlags 8
    pattern ImGuiColorEditFlags_NoSmallPreview :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoSmallPreview = ImGuiColorEditFlags 16
    pattern ImGuiColorEditFlags_NoInputs :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoInputs = ImGuiColorEditFlags 32
    pattern ImGuiColorEditFlags_NoTooltip :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoTooltip = ImGuiColorEditFlags 64
    pattern ImGuiColorEditFlags_NoLabel :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoLabel = ImGuiColorEditFlags 128
    pattern ImGuiColorEditFlags_NoSidePreview :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoSidePreview = ImGuiColorEditFlags 256
    pattern ImGuiColorEditFlags_NoDragDrop :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoDragDrop = ImGuiColorEditFlags 512
    pattern ImGuiColorEditFlags_NoBorder :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_NoBorder = ImGuiColorEditFlags 1024
    pattern ImGuiColorEditFlags_AlphaBar :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_AlphaBar = ImGuiColorEditFlags 65536
    pattern ImGuiColorEditFlags_AlphaPreview :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_AlphaPreview = ImGuiColorEditFlags 131072
    pattern ImGuiColorEditFlags_AlphaPreviewHalf :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_AlphaPreviewHalf = ImGuiColorEditFlags 262144
    pattern ImGuiColorEditFlags_HDR :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_HDR = ImGuiColorEditFlags 524288
    pattern ImGuiColorEditFlags_DisplayRGB :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_DisplayRGB = ImGuiColorEditFlags 1048576
    pattern ImGuiColorEditFlags_DisplayHSV :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_DisplayHSV = ImGuiColorEditFlags 2097152
    pattern ImGuiColorEditFlags_DisplayHex :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_DisplayHex = ImGuiColorEditFlags 4194304
    pattern ImGuiColorEditFlags_Uint8 :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_Uint8 = ImGuiColorEditFlags 8388608
    pattern ImGuiColorEditFlags_Float :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_Float = ImGuiColorEditFlags 16777216
    pattern ImGuiColorEditFlags_PickerHueBar :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_PickerHueBar = ImGuiColorEditFlags 33554432
    pattern ImGuiColorEditFlags_PickerHueWheel :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_PickerHueWheel = ImGuiColorEditFlags 67108864
    pattern ImGuiColorEditFlags_InputRGB :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_InputRGB = ImGuiColorEditFlags 134217728
    pattern ImGuiColorEditFlags_InputHSV :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags_InputHSV = ImGuiColorEditFlags 268435456
    pattern ImGuiColorEditFlags__OptionsDefault :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags__OptionsDefault = ImGuiColorEditFlags 177209344
    pattern ImGuiColorEditFlags__DisplayMask :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags__DisplayMask = ImGuiColorEditFlags 7340032
    pattern ImGuiColorEditFlags__DataTypeMask :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags__DataTypeMask = ImGuiColorEditFlags 25165824
    pattern ImGuiColorEditFlags__PickerMask :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags__PickerMask = ImGuiColorEditFlags 100663296
    pattern ImGuiColorEditFlags__InputMask :: ImGuiColorEditFlags
    pattern ImGuiColorEditFlags__InputMask = ImGuiColorEditFlags 402653184
    newtype ImGuiSliderFlags
      = ImGuiSliderFlags Foreign.C.Types.CInt
      deriving newtype (Eq, Ord, Data.Bits.Bits)
    pattern ImGuiSliderFlags_None :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_None = ImGuiSliderFlags 0
    pattern ImGuiSliderFlags_AlwaysClamp :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_AlwaysClamp = ImGuiSliderFlags 16
    pattern ImGuiSliderFlags_Logarithmic :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_Logarithmic = ImGuiSliderFlags 32
    pattern ImGuiSliderFlags_NoRoundToFormat :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_NoRoundToFormat = ImGuiSliderFlags 64
    pattern ImGuiSliderFlags_NoInput :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_NoInput = ImGuiSliderFlags 128
    pattern ImGuiSliderFlags_InvalidMask_ :: ImGuiSliderFlags
    pattern ImGuiSliderFlags_InvalidMask_ = ImGuiSliderFlags 1879048207
    newtype ImGuiMouseButton
      = ImGuiMouseButton Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiMouseButton_Left :: ImGuiMouseButton
    pattern ImGuiMouseButton_Left = ImGuiMouseButton 0
    pattern ImGuiMouseButton_Right :: ImGuiMouseButton
    pattern ImGuiMouseButton_Right = ImGuiMouseButton 1
    pattern ImGuiMouseButton_Middle :: ImGuiMouseButton
    pattern ImGuiMouseButton_Middle = ImGuiMouseButton 2
    newtype ImGuiMouseCursor
      = ImGuiMouseCursor Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiMouseCursor_None :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_None = ImGuiMouseCursor (-1)
    pattern ImGuiMouseCursor_Arrow :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_Arrow = ImGuiMouseCursor 0
    pattern ImGuiMouseCursor_TextInput :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_TextInput = ImGuiMouseCursor 1
    pattern ImGuiMouseCursor_ResizeAll :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_ResizeAll = ImGuiMouseCursor 2
    pattern ImGuiMouseCursor_ResizeNS :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_ResizeNS = ImGuiMouseCursor 3
    pattern ImGuiMouseCursor_ResizeEW :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_ResizeEW = ImGuiMouseCursor 4
    pattern ImGuiMouseCursor_ResizeNESW :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_ResizeNESW = ImGuiMouseCursor 5
    pattern ImGuiMouseCursor_ResizeNWSE :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_ResizeNWSE = ImGuiMouseCursor 6
    pattern ImGuiMouseCursor_Hand :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_Hand = ImGuiMouseCursor 7
    pattern ImGuiMouseCursor_NotAllowed :: ImGuiMouseCursor
    pattern ImGuiMouseCursor_NotAllowed = ImGuiMouseCursor 8
    newtype ImGuiCond
      = ImGuiCond Foreign.C.Types.CInt
      deriving newtype (Eq, Ord)
    pattern ImGuiCond_None :: ImGuiCond
    pattern ImGuiCond_None = ImGuiCond 0
    pattern ImGuiCond_Always :: ImGuiCond
    pattern ImGuiCond_Always = ImGuiCond 1
    pattern ImGuiCond_Once :: ImGuiCond
    pattern ImGuiCond_Once = ImGuiCond 2
    pattern ImGuiCond_FirstUseEver :: ImGuiCond
    pattern ImGuiCond_FirstUseEver = ImGuiCond 4
    pattern ImGuiCond_Appearing :: ImGuiCond
    pattern ImGuiCond_Appearing = ImGuiCond 8
```

<details>

sheaf commented:
With [Luke Lau's patch](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3330) I'm able to add Haddock documentation to the Template Haskell splices. Just running `cabal haddock` generates the expected documentation, no need for a custom setup like @ocharles feared.

The documentation text still needs to be correctly processed for formatting/cross-linking. Here's a snapshot (can't embed HTML on GitHub):

![context_docs](https://user-images.githubusercontent.com/1297748/106373495-5baa9d00-637a-11eb-85cf-59bc4ea43186.png)

The only limitation I've run into is that I'm unable to bundle the pattern synonyms as there's no way of writing module exports with TH.

sheaf commented:
This passes CI now.

ocharles commented on Main.hs:
```diff
@@ -70,7 +70,7 @@ loop w checked color slider r = do
     False -> return ()
 
-  sameLine >> arrowButton "Arrow" ImGuiDirUp
+  sameLine >> arrowButton "Arrow" ImGuiDir_Up
```
A bit of a shame - do you think we should drop `_`, or have naming as exactly close to C as possible?

sheaf replied:
My reasoning was that there are fields with double underscores (e.g. `ImGuiColorEditFlags__OptionsDefault`), and I thought it would be a bit confusing to now have those with single underscores and everything else with no underscores. I don't really mind though, if you prefer removing the first underscore I can do that.

ocharles replied:
Ok, let's keep this how it is - that's a great point and I think it'll be more confusing if we change it.

ocharles commented:
I tried this out and it works perfectly! I'm happy to merge this when you are. If you want to try more things first that's fine, just give me a heads up when you're ready.

TristanCacqueray commented on src/DearImGui/Enums.hs:
```diff
@@ -0,0 +1,14 @@
+
+--------------------------------------------------------------------------------
+
+declareEnumerations
```
Is it possible to print what declareEnumerations output? If so, would it be ok to add some instructions for contributor not familiar with template haskell?

sheaf replied:
`haskell-language-server` has a "splice" plugin which can be used to see the contents of the splice. You can also pass the `-ddump-splices` flag to GHC, which will output the result. However in this case I think it would suffice to read the documentation generated by haddock.

sheaf commented:
I've changed the generator to use `Lift` instead of faffing around with `TVar`s, much better now, thanks for the suggestion @ocharles.

sheaf commented:
- Added `FiniteEnum` typeclass and instances, to handle the `COUNT` patterns, 
- Added `Storable` instances, so that these types can be used with `inline-c`.

This should be OK to merge.

ocharles commented:
Thanks for this, @sheaf! While there might be small things we want to tune, I think this PR is definitely in a fine state to be merged as-is so I'm gonna bring it in. Thanks once again!
